Draft/RFC0001-PowerShell-Based-LCM.md

```diff
@@ -0,0 +1,92 @@
+---
+RFC: 0001
+Author: Ryan Bartram
+Status: Draft
+Version: 1.0
+Area: Engine
+Comments Due: 14/11/2019
+Plan to implement: <Yes | No>
+---
+
+# PowerShell Based LCM
+
+The Local Configuration Manager (LCM) was a windows component made available with the
+release of the WMF 4. It's the engine behind Desired State Configuration (DSC) and is used
+predominently as a Windows based configuration management tool.
+
+The current implementation is based on WMI/OMI and receives MOF files either from a Pull Server
+or by use of the cmdlets `Start-DscConfiguration` or `Publish-DscConfiguration`.
+
+The LCM can be configured to pull its configuration on a recurring basis and implements security
+by verifying signatures of MOFs or modules before they are downloaded.
+
+The LCM has 3 application modes `ApplyOnly`, `ApplyAndMonitor` and `ApplyAndAutoCorrect`. When
+the configuration is applied, it is applied in a serial fashion i.e. one resource after another.
+There is a dependency model defined within the configuration which enforces the order with which
+resources are deployed.
+
+## Motivation
+
+As a user, I want to replace the existing WMI/OMI based LCM on Windows and Linux with cross
+platform version that can perform all of the tasks of the existing LCM
+plus a lot lot more. So I can utilise my existing DSC knowledge to manage infrastructure
+all over my organisation. Onprem and in the cloud.
+
+I see this being the new Terraform, Polumi etc
+
+## Specification
+
+### User Experience
+
+```PowerShell
+# Command to generate new DSC MOF file
+Build-DscConfiguration
```

Comment:
We should make sure we decouple the DSC Compilation from the invocation.
The Compilation and the `Configuration` keyword are parts of PowerShell and partly from the PSDesiredStateConfiguration.psm1 module.
I'm pushing for the latter to be Open Sourced, and for some work to be done by the Config team to allow decoupling compilation from Format (so that we could access in-memory representation of the resource graph).
We'll see where that goes, but Open Sourcing the PSDesiredStateConfiguration module should be possible in the mid-future (~6months kind of timeframe)

Reply:
you are right here. That should be another RFC. My thinking here was because I wanted to adjust the schema that we would need a new compiler. But with hinsight, supporting existing MOFs is a must. Doesn't mean we can't support newer features but thats out of scope for the invocation.